// @celo/rainbowkit-celo/wallets/celo-wallet.js
import { getWalletConnectConnector } from "@rainbow-me/rainbowkit";
import { Alfajores, Baklava, Celo } from "@celo/rainbowkit-celo/chains";
export const CeloWallet = ({ chains = [Alfajores, Baklava, Celo], }) => ({
    id: "celo-wallet",
    name: "Celo Wallet",
    iconUrl: "https://rainbowkit-with-celo.vercel.app/icons/mono.svg",
    iconBackground: "#FFF",
    createConnector: () => {
        const connector = getWalletConnectConnector({
            chains,
        });
        return {
            connector,
            mobile: {
                getUri: async () => {
                    const { uri } = (await connector.getProvider()).connector;
                    return uri;
                },
            },
            desktop: {
                getUri: async () => {
                    const { uri } = (await connector.getProvider()).connector;
                    return `celowallet://wc?uri=${encodeURIComponent(uri)}`;
                },
            },
        };
    },
});
export default CeloWallet;
