// redstone-protocol/dist/src/common/redstone-constants.js
"use strict";
// Abbreviations:
// * BS = Byte size
Object.defineProperty(exports, "__esModule", { value: true });
exports.DATA_FEED_ID_BS = exports.SIGNATURE_BS = exports.REDSTONE_MARKER_BS = exports.REDSTONE_MARKER_HEX = exports.UNSGINED_METADATA_BYTE_SIZE_BS = exports.DATA_PACKAGES_COUNT_BS = exports.DEFAULT_NUM_VALUE_DECIMALS = exports.DEFAULT_NUM_VALUE_BS = exports.DATA_POINT_VALUE_BYTE_SIZE_BS = exports.DATA_POINTS_COUNT_BS = exports.TIMESTAMP_BS = void 0;
// Number of bytes reserved to store timestamp
exports.TIMESTAMP_BS = 6;
// Number of bytes reserved to store the number of data points
exports.DATA_POINTS_COUNT_BS = 3;
// Number of bytes reserved to store datapoints byte size
exports.DATA_POINT_VALUE_BYTE_SIZE_BS = 4;
// Default value byte size for numeric values
exports.DEFAULT_NUM_VALUE_BS = 32;
// Default precision for numeric values
exports.DEFAULT_NUM_VALUE_DECIMALS = 8;
// Number of bytes reserved for data packages count
exports.DATA_PACKAGES_COUNT_BS = 2;
// Number of bytes reserved for unsigned metadata byte size
exports.UNSGINED_METADATA_BYTE_SIZE_BS = 3;
// RedStone marker, which will be appended in the end of each transaction
exports.REDSTONE_MARKER_HEX = "0x000002ed57011e0000";
// Byte size of RedStone marker
// we subtract 1 because of the 0x prefix
exports.REDSTONE_MARKER_BS = exports.REDSTONE_MARKER_HEX.length / 2 - 1;
// Byte size of signatures
exports.SIGNATURE_BS = 65;
// Byte size of data feed id
exports.DATA_FEED_ID_BS = 32;
